log4j-core/src/main/java/org/apache/logging/log4j/core/impl/MementoLogEvent.java

```diff
@@ -72,50 +73,37 @@ public MementoLogEvent(final LogEvent event) {
         if (instant.getEpochMillisecond() == 0 && message instanceof TimestampMessage) {
             instant.initFromEpochMilli(((TimestampMessage) message).getTimestamp(), 0);
         }
-        contextData = memento(event.getContextData());
+        contextData = mementoOfContextData(event.getContextData());
         contextStack = event.getContextStack();
-        source = includeLocation ? event.getSource() : null;
+        source = includeLocation ? event.getSource() : event.peekSource();
         threadName = event.getThreadName();
         threadId = event.getThreadId();
         threadPriority = event.getThreadPriority();
         thrown = event.getThrown();
         thrownProxy = event.getThrownProxy();
     }
 
-    public MementoLogEvent(final LogEvent event, final boolean includeLocation) {
-        loggerFqcn = event.getLoggerFqcn();
-        loggerName = event.getLoggerName();
-        instant.initFrom(event.getInstant());
-        nanoTime = event.getNanoTime();
-        level = event.getLevel();
-        marker = event.getMarker();
-        locationRequired = includeLocation;
-        endOfBatch = event.isEndOfBatch();
-        message = mementoOfMessage(event);
-        if (instant.getEpochMillisecond() == 0 && message instanceof TimestampMessage) {
-            instant.initFromEpochMilli(((TimestampMessage) message).getTimestamp(), 0);
+    private static ReadOnlyStringMap mementoOfContextData(final ReadOnlyStringMap readOnlyMap) {
+        if (readOnlyMap instanceof final StringMap stringMap && !stringMap.isFrozen()) {
+            final StringMap data = ContextDataFactory.createContextData(readOnlyMap);
+            data.freeze();
+            return data;
         }
-        contextData = memento(event.getContextData());
-        contextStack = event.getContextStack();
-        source = includeLocation ? event.getSource() : null;
-        threadName = event.getThreadName();
-        threadId = event.getThreadId();
-        threadPriority = event.getThreadPriority();
-        thrown = event.getThrown();
-        thrownProxy = event.getThrownProxy();
+        // otherwise immutable
+        return readOnlyMap;
     }
 
-    @Override
-    public LogEvent toImmutable() {
-        return this;
+    private static Message mementoOfMessage(final LogEvent event) {
+        final Message message = event.getMessage();
+        if (message instanceof LoggerNameAwareMessage) {
+            ((LoggerNameAwareMessage) message).setLoggerName(event.getLoggerName());
+        }
```

Comment:
_[I know this code block is not actually changed.]_ I am not able to follow why this is the right place to do this. Shouldn't this already be done while the message gets populated?

Reply:
Same as above: let's deal with it in another PR.